File: server/backend/src/backend/api/models/download_profile_podcast.py
from __future__ import annotations

from datetime import datetime

from pydantic import Field

from backend.api.models.base import RequestBase, ResponseBase


# ---------- Strict input (create/update) ----------
class _DownloadProfilePodcastAPIBaseIn(RequestBase):
    """Fields for requests: validate here (constraints allowed)."""

    enable_profile: bool
    download_with_countdown: bool
    redownload_final: bool
    download_days_in_past: int = Field(ge=0)
    delete_older_episodes: bool


class DownloadProfilePodcastAPICreateBundle(_DownloadProfilePodcastAPIBaseIn):
    """Request body for creating a download profile for podcasts while bundleing it with a show and media profile."""
    pass


class DownloadProfilePodcastAPICreate(_DownloadProfilePodcastAPIBaseIn):
    """Request body for creating a download profile for podcasts."""

    show_id: int
    media_profile_id: int


class DownloadProfilePodcastAPIUpdate(_DownloadProfilePodcastAPIBaseIn):
    """Request body for updating a download profile for podcasts."""

    media_profile_id: int


# ---------- Lenient output (read) ----------
class _DownloadProfilePodcastAPIBaseOut(ResponseBase):
    """Fields for responses: no validators, no constraints."""

    id: int
    show_id: int
    media_profile_id: int
    enable_profile: bool
    download_with_countdown: bool
    redownload_final: bool
    download_days_in_past: int
    delete_older_episodes: bool


class DownloadProfilePodcastAPIRead(_DownloadProfilePodcastAPIBaseOut):
    """Response body for a download profile for podcasts."""

    created_at: datetime
    updated_at: datetime

File: ui/src/types/schemas/download_profile_podcast.ts
import {z} from 'zod';


// ---------- Strict request (create/update) ----------
const DownloadProfilePodcastBaseSchema = z.object({
    mediaProfileId: z.int(),
    enableProfile: z.boolean().default(true),
    downloadWithCountdown: z.boolean().default(false),
    redownloadFinal: z.boolean().default(true),
    downloadDaysInPast: z.int().min(0).default(180),
    deleteOlderEpisodes: z.boolean().default(true),
})

export const DownloadProfilePodcastCreateSchema = DownloadProfilePodcastBaseSchema.extend({
    showId: z.int(),
})
export type DownloadProfilePodcastCreateIn = z.input<typeof DownloadProfilePodcastCreateSchema>
export type DownloadProfilePodcastCreateOut = z.output<typeof DownloadProfilePodcastCreateSchema>


export const DownloadProfilePodcastUpdateSchema = DownloadProfilePodcastBaseSchema.extend({
})
export type DownloadProfilePodcastUpdateIn = z.input<typeof DownloadProfilePodcastUpdateSchema>
export type DownloadProfilePodcastUpdateOut = z.output<typeof DownloadProfilePodcastUpdateSchema>


// ------------ Lenient response (read) ------------
export const DownloadProfilePodcastReadSchema = z.looseObject({
    id: z.int(),
    showId: z.int(),
    mediaProfileId: z.int().optional(),
    enableProfile: z.boolean(),
    downloadWithCountdown: z.boolean(),
    redownloadFinal: z.boolean(),
    downloadDaysInPast: z.int(),
    deleteOlderEpisodes: z.boolean(),
    createdAt: z.iso.datetime().transform((s) => new Date(s)),
    updatedAt: z.iso.datetime().transform((s) => new Date(s)),
})
export type DownloadProfilePodcastRead = z.infer<typeof DownloadProfilePodcastReadSchema>